flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/collect/CollectSinkFunction.java

```diff
@@ -0,0 +1,381 @@
+		private DataOutputViewStreamWrapper outStream;
+
+		private ServerThread() throws Exception {
+			this.serverSocket = new ServerSocket(0, 0, getBindAddress());
+			this.running = true;
+		}
+
+		@Override
+		public void run() {
+			while (running) {
+				try {
+					if (connection == null) {
+						// waiting for coordinator to connect
+						connection = serverSocket.accept();
+						inStream = new DataInputViewStreamWrapper(this.connection.getInputStream());
+						outStream = new DataOutputViewStreamWrapper(this.connection.getOutputStream());
+						LOG.info("Coordinator connection received");
+					}
+
+					CollectCoordinationRequest request = new CollectCoordinationRequest(inStream);
+					String requestVersion = request.getVersion();
+					long requestOffset = request.getOffset();
+					if (LOG.isDebugEnabled()) {
+						LOG.debug(
+							"Request received, version = " + requestVersion + ", offset = " + requestOffset);
+						LOG.debug(
+							"Expecting version = " + version + ", firstBufferedOffset = " + firstBufferedResultOffset);
+					}
+
+					if (!version.equals(requestVersion) || requestOffset < firstBufferedResultOffset) {
+						// invalid request
+						if (LOG.isDebugEnabled()) {
+							LOG.debug("Invalid request");
+						}
+						sendBufferedResults(0, 0);
+						continue;
+					}
+
+					// valid request, sending out results
+					try {
+						bufferedResultsLock.lock();
+
+						int oldSize = bufferedResults.size();
+						int start = Math.min((int) (requestOffset - firstBufferedResultOffset), oldSize);
+						int end = Math.min(start + maxResultsPerBatch, oldSize);
+
+						if (LOG.isDebugEnabled()) {
+							LOG.debug("Sending back " + (end - start) + " results");
+						}
+						sendBufferedResults(start, end - start);
+
+						if (oldSize >= maxResultsBuffered && bufferedResults.size() < maxResultsBuffered) {
+							bufferNotFullCondition.signal();
+						}
+					} finally {
+						bufferedResultsLock.unlock();
+					}
+				} catch (IOException e) {
+					// IOException occurs, just close current connection
+					// client will come with the same offset if it needs the same batch of results
+					if (LOG.isDebugEnabled()) {
+						LOG.debug("Collect sink server encounters an IOException", e);
+					}
+					closeCurrentConnection();
+				}
+			}
+		}
+
+		private void close() {
+			running = false;
+			closeCurrentConnection();
+			closeServer();
+		}
+
+		private InetSocketAddress getServerSocketAddress() {
+			RuntimeContext context = getRuntimeContext();
+			Preconditions.checkState(
+				context instanceof StreamingRuntimeContext,
+				"CollectSinkFunction can only be used in StreamTask");
+			StreamingRuntimeContext streamingContext = (StreamingRuntimeContext) context;
+			String taskManagerAddress = streamingContext.getTaskManagerRuntimeInfo().getTaskManagerExternalAddress();
+			return new InetSocketAddress(taskManagerAddress, serverSocket.getLocalPort());
+		}
+
+		private InetAddress getBindAddress() {
```

Comment:
useless method?

Reply:
It is used when creating the socket server. See line 241.

Reply:
isn't this address the same as what you get from `StreamingRuntimeContext`?

Reply:
No. Binding address is different from external address. Binding address is an internal IP address which means that the socket only listens to the requests coming from this network interface (the default binding address is 0.0.0.0 which is a wildcard address, indicating that the socket listens to requests coming from all network interfaces), while external address is an external IP address.